src/index.js

```diff
@@ -24,275 +26,260 @@ function init () {
+        log(`${_fullname(newSocket)} removed connections`)
+        logData(connections)
+      })
+      .on('replaceConnections', (connections) => {
+        connections = filterNewConnections(connections)
 
-        joinChannel(socket, socket.channelName)
+        channelGraph().clearConnections()
+        channelGraph().addConnections(connections)
+        sendToChannel('connections', channelGraph().listConnections())
 
-        socket.clientDescription.member = socket.clientDescription.name // legacy
-        io.to(socket.channelName).emit('new-member', socket.clientDescription) // legacy
-        io.to(socket.channelName).emit('newClient', socket.clientDescription)
+        log(`${_fullname(newSocket)} replaced connections`)
+        logData(connections)
       })
-      // TODO: filter by channel
-      .on('addConnections', (data) => addConnections(data, socket))
-      // TODO: filter by channel
-      .on('removeConnections', (data) => removeConnections(data, socket))
-      // TODO: filter by channel
       .on('getConnections', (data) => {
-        io.to(socket.id).emit('connections', connections[socket.channelName])
+        sendBack('connections', channelGraph().listConnections())
+      })
+
+    newSocket
+      .on('removeClients', (clientNames) => {
+        for (const name of clientNames) {
+          channelGraph().removeClient(name)
+        }
+        sendToChannel('clients', channelGraph().listClients())
       })
-      // TODO: filter by channel
       .on('getClients', (data) => {
-        io.to(socket.id).emit('clients', getClients(socket))
+        sendBack('clients', channelGraph().listClients())
       })
-      .on('replaceConnections', (data) => {
-        data = objectify(data)
-        if (data) {
-          if (Array.isArray(data)) {
-            connections[socket.channelName] = data
-          } else {
-            connections[socket.channelName] = [data]
-          }
-          settings.verbose && log(`${fullname(socket)} replaced connections`)
-          !settings.semiverbose && jsonColorz(data)
-          // remove duplicated
-          connections[socket.channelName] = _.uniqWith(connections[socket.channelName], _.isEqual)
-        }
+
+    newSocket
+      .on('saveGraph', (data) => {
+        saveGraph(newSocket.channelName)
       })
       .on('*', ({ data }) => {
         let [eventName, args] = data
+        const { _to } = args
+
+        if (reservedEvents.indexOf(eventName) > -1) {
+          return
+        }
 
-        if (reservedEvents.indexOf(eventName) > -1) return
+        if (settings.showdashboard) {
+          dashboard.observeEvent(infos, eventName, newSocket.channelName)
+        }
 
-        observeEvent(eventName, socket.channelName)
+        log(`${_fullname(newSocket)} emitted event "${eventName}"`)
+        logData(data)
 
-        if (typeof args !== 'object') {
-          args = {data: args}
-          args.altered = true
+        if (args.altered) {
```

Comment:
keep the code adding the altered flag as before, and add a comment saying it is legacy